### PR TITLE
[MOFONETPROFILES-84] extract neo4j-http gem from doximity

### PR DESCRIPTION
**Pull request by rstawarz**
Initial commit of gem being extracted from Doximity

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+      - image: cimg/ruby:2.7
+        environment:
+          - BUNDLE_VERSION: "~> 1.17"
+          - NEO4J_URL: "http://127.0.0.1:7474"
+      - image: neo4j:4.3.9-enterprise
+        environment:
+          NEO4J_ACCEPT_LICENSE_AGREEMENT: "yes"
+          NEO4J_AUTH: "none"
+          NEO4J_dbms_allow__upgrade: "true"
+          NEO4J_dbms_jvm_additional: "-Xms256m -Xmx256m"
```
Comment (bsimpson) [inline .circleci/config.yml:19]: Do we want this set at 256? I don't know what we had before but this won't let any memory grow

File: README.md [hunk 2]
```diff
@@ -1,2 +1,119 @@
...
+
+is equivalent to:
+
+```
+config = Neo4j::Http.config
+cypher_client = Neo4j::Http::CypherClient.new(config)
+node_client = Neo4j::Http::NodeClient.new(cypher_client)
+```
+
+to connect to a different Neo4j datbase, you can create a custom configuration like:
```
Comment (bsimpson) [inline README.md:85]: datbase -> database

File: README.md [hunk 3]
```diff
@@ -1,2 +1,119 @@
...
+You can also run `bin/console` for an interactive prompt that will allow you to experiment.
+
+To install this gem onto your local machine, run `bundle exec rake install`.
+
+To release a new version, update the version number in `version.rb`, issue a pull request, and once merged with passing CI, the new gem version will be pushed automatically.
+
+## Contributing
+
+1. Fork it
+2. Create your feature branch (`git checkout -b my-new-feature`)
```
Comment (bsimpson) [inline README.md:111]: `git switch -c my-new-feature` is the 2.6+ syntax

File: README.md [hunk 4]
```diff
@@ -1,2 +1,119 @@
...
+
+## Versioning
+
+This project follows [semantic versioning](https://semver.org).
+
+## Development
+
+After checking out the repo, run `bin/setup` to install dependencies.
+
+To run specs, you'll need a running neo4j instance available at `localhst:7474`.  If you have Docker installed, this is easily done by using the provided [docker-file](https://github.com/doximity/neo4j-http/blob/master/docker-compose.yml) - simply run `docker-compose up` within the project directory, and once running, you can then, run `rake spec` to run the tests in another terminal.
```
Comment (bsimpson) [inline README.md:100]: localhst -> localhost

File: README.md [hunk 5]
```diff
@@ -1,2 +1,119 @@
-# neo4j-http
-A simple ruby HTTP client for Neo4j
+# Neo4j::Http
+
+The `Neo4j::Http` gem provides `Neo4j::Http::Client` as a thin wrapper around the [Neo4j HTTP API](https://neo4j.com/docs/http-api/current/).
```
Comment (bsimpson) [inline README.md:3]: I wonder if it is worth explaining who needs this, and what the difference is between the HTTP API and the deprecated REST API? Maybe even highlight that this isn't a native driver that requires OS dependencies? We know the benefits but I think a README being the first thing users see is where we can "sell" it

File: README.md [hunk 6]
```diff
@@ -1,2 +1,119 @@
...
+
+    $ bundle install
+
+Or install it yourself as:
+
+    $ gem install neo4j-http
+
+## Configuration
+
+The client is configured by default via a set of environment variables from `[Neo4j::Http::Configuration](https://github.com/doximity/neo4j-http/blob/master/lib/neo4j/http/configuration.rb)`:
```
Comment (bsimpson) [inline README.md:23]: You won't be able to render the link inside the backticks. It will just show a literal `[xxx](xxx)`

File: README.md [hunk 7]
```diff
@@ -1,2 +1,119 @@
...
+
+* `NEO4J_URL`  - The base URL to connect to Neo4j at
+* `NEO4J_USER` - The user name to use when authenticating to neo4j
+* `NEO4J_PASSWORD` - The password of the user to be used for authentication
+* `NEO4J_DATABASE` - The database name to be used when connectiong.  By default this will be nil and the path used for connecting to Neo4j wil be `/db/data/transaction/commit` to make it compliant with v3.5 of neo4j
```
Comment (bsimpson) [inline README.md:28]: connectiong -> connecting

File: README.md [hunk 8]
```diff
@@ -1,2 +1,119 @@
...
+* `NEO4J_HTTP_USER_AGENT` - The user agent name provided in the request - defaults to `Ruby Neo4j Http Client`
+* `NEO4J_REQUEST_TIMEOUT_IN_SECONDS` - The number of seconds for the http request to time out if provided, defaults to nil
+
+These configuration values can also be set during intialization like:
```
Comment (bsimpson) [inline README.md:32]: intialization -> initialization

File: README.md [hunk 9]
```diff
@@ -1,2 +1,119 @@
...
+
+```
+Neo4j::Http.configure do |config|
+  config.request_timeout_in_secods = 42
```
Comment (bsimpson) [inline README.md:36]: request_timeout_in_secods -> request_timeout_in_seconds

File: README.md [hunk 10]
```diff
@@ -1,2 +1,119 @@
...
+end
+```
+
+## Usage
+
+The core interface can be directly accessed on `Neo4::Http::Client`:
+
+```
+# Execute arbitrary cypher commands
```
Comment (bsimpson) [inline README.md:45]: I would probably break these up into subheadings. Right now it renders as a big wall of text

File: README.md [hunk 11]
```diff
@@ -1,2 +1,119 @@
...
+Neo4j::Http::Client.find_relationship(relationship: relationship, from: user1, to: user2)
+
+# Delete the relationship if it exists
+Neo4j::Http::Client.delete_relationship(relationship: relationship, from: user1, to: user2)
+```
+
+Each of the methods exposed on `Neo4j::Http::Client` above are provided by instances of each of the following adapters:
+* `Neo4j::Http::CypherClient` - provides an `execute_cypher` method which sends raw cypher commands to neo4j
+* `Neo4j::Http::NodeClient` - provides a higher level API for upserting and deleting Nodes
+* `Neo4j::Http::RelationshipClient` - provides a higher levle API for upserting and deleting Relationships
```
Comment (bsimpson) [inline README.md:71]: levle -> level

File: README.md (hunk 10 above)
Comment (bsimpson) [reply README.md:45]: Also lets throw a hint to the highlighter e.g. ```ruby

File: README.md [hunk 12]
```diff
@@ -1,2 +1,119 @@
...
+
+## Contributing
+
+1. Fork it
+2. Create your feature branch (`git checkout -b my-new-feature`)
+3. Commit your changes (`git commit -am 'Add some feature'`)
+4. Push to the branch (`git push origin my-new-feature`)
+5. Create a new Pull Request
+6. Sign the CLA if you haven't yet. See [CONTRIBUTING.md](https://github.com/doximity/neo4j-http/blob/master/CONTRIBUTING.md)
+
```
Comment (bsimpson) [inline README.md:116]: Can we add a section here about how to run any tests?

File: .circleci/config.yml (hunk 1 above)
Comment (rstawarz) [reply .circleci/config.yml:19]: I think i copied this from what we have in the doximity circleci config

File: README.md (hunk 12 above)
Comment (rstawarz) [reply README.md:116]: That's included here: https://github.com/doximity/neo4j-http/pull/1/files/00e99e7340caf09e933c4a3924ce8f0ab2118848#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R100

File: README.md [hunk 13]
```diff
@@ -1,6 +1,9 @@
 # Neo4j::Http
 
-The `Neo4j::Http` gem provides `Neo4j::Http::Client` as a thin wrapper around the [Neo4j HTTP API](https://neo4j.com/docs/http-api/current/).
+The `Neo4j::Http` gem provides `a thin wrapper around the [Neo4j HTTP API](https://neo4j.com/docs/http-api/current/) (not the legacy REST api which was removed in 4.0). It works with Neo4j 3.5 through the latest release (at the time of this writing is 4.4)
```
Comment (bsimpson) [inline README.md:3]: Remove "`" in "`a thin wrapper"

File: README.md [hunk 14]
```diff
@@ -18,13 +21,96 @@ Or install it yourself as:
...
+
+```
+config = Neo4j::Http.config
+cypher_client = Neo4j::Http::CypherClient.new(config)
+node_client = Neo4j::Http::NodeClient.new(cypher_client)
+```
+
+to connect to a different Neo4j database, you can create a custom configuration like:
+```
+config = Neo4j::Http::Configuration.new({ datbase_name: 'test' })
```
Comment (bsimpson) [inline README.md:98]: datbase_name -> database_name

File: README.md [hunk 15]
```diff
@@ -33,7 +119,7 @@ To release a new version, update the version number in `version.rb`, issue a pul
 ## Contributing
 
 1. Fork it
-2. Create your feature branch (`git checkout -b my-new-feature`)
+2. Create your feature branch ( git switch -c my-new-feature`)
```
Comment (bsimpson) [inline README.md:122]: Add the matching "`" here

**Review by joe-armstrong — APPROVED** [review]
LGTM from a security perspective.

**Comment by anovadox** [comment]
> @rstawarz so the one thing I’m a bit worried about here is the circleci context. We have the artifact_publishing context specified, which I assume has some secrets like nexus credentials and rubygems credentials.
> 
> But I think maybe there is extra work required to prevent people from just forking the repo and then running arbitrary commands in circleci, which will have those env vars available via the context.
> 
> Random related blog post I found: https://circleci.com/blog/managing-secrets-when-you-have-pull-requests-from-outside-contributors/
> 
> Can we maybe get someone from @doximity/infra_automation to weigh in on this? Is it safe to have this context available to circle, even in a public repo?
> 
> Edit:
> 
> Looks like we don't have this setting enabled for this repo. I guess it's off by default, which makes me feel better.
> 
> It does make me wonder whether there's some failsafe "never allow passing contexts to public forks for my entire org" button we could click to prevent this from accidentally happening.
> 
> Screen Shot 2022-04-13 at 10 25 43 AM

I also sort of hate that this toggle is just _available_ with no configuration on our end, and more than that: no papertrail if it is enabled.

But it is off by default.

File: README.md [hunk 16]
```diff
@@ -18,13 +21,96 @@ Or install it yourself as:
...
+end
+```
+
+## Usage
+
+The core interface can be directly accessed on `Neo4::Http::Client` -
+
+### Execute arbitrary cypher commands
+```ruby
+Neo4j::Http::Client.execute_cypher('MATCH n:User{id: $id} return n LIMIT 25', id: 42)
```
Comment (bsimpson) [inline README.md:49]: I tried this command and received a syntax error saying "Parentheses are required to identify nodes in patterns". This should be:

```
Neo4j::Http::Client.execute_cypher('MATCH (n:User{id: $id}) return n LIMIT 25', id: 42)
```

**Comment by rstawarz** [comment]
@jeffgran - Given @anovadox comment - are you ok with this PR?

**Review by jeffgran — APPROVED** [review]
@rstawarz yes, please go ahead.

**Pull request merged**